cmd/pebble/cmd_plan.go

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) 2021 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+
+	"github.com/jessevdk/go-flags"
+
+	"github.com/canonical/pebble/client"
+)
+
+type cmdMerge struct {
+	clientMixin
+	Positional struct {
+		LayerPath string `positional-arg-name:"<layer-path>" required:"1"`
+	} `positional-args:"yes"`
+}
+
+var shortMergeHelp = "Dynamically merge a layer on top of the plan's layers"
```

Comment:
I've spent some time here thinking about how we organize things in a way that, as usual, is fast for us to support right now, while leaving the road open for us to implement all the things we want in the near future. I've struggled a bit for two reasons:

1) "merge" seems to mean something else here vs. the internal "override" directive that exists in services. The best way to see that conflict is to think about the fact we have "merge" vs. "replace", in that directive. But if we have "merge" vs. "replace", we then have "merge" with "merge", or "merge" with "replace". Ugh. Terminology is hard.

2) Second reason is that this seems like a more fundamental operation: add a layer. There will likely be several different ways of adding layers: combining them, not combining them, dynamically, statically, persisting, etc. We don't want independent commands for all of those cases. Instead, it'd be much nicer to have one _solid_ command that has the most obvious default behavior, and then flags to customize the remaining options.

So, here is an idea: we already have a word for the idea of getting two different layers and merging them together into a single layer: combining. So, what if we changed both the API and the CLI here so that we spell out what's going on here a bit more precisely?  In other words, this operation would become:
```
$ pebble add --combine --dynamic <layer path>
```
Which is translated into an operation with `action: "add", "combine": true, "dynamic": true`.

In the near future, we'll fill the gaps and introduce the opportunity to not have these options set. For now, we can simply error out on the server side if they aren't set (`error: layer must be combined and dynamic - other options coming soon`), so that this PR can land sooner rather than later. The CLI and client can already support these options properly by forwarding them over, since that's trivial.

How does that sound?

Reply:
Thanks. I like the idea of "add" with different modes / sub-actions. However, two things I'm not sure about:

1) I think it's a bit annoying to have to specify `--combine --dynamic` when at least for now that's all we have, and I think those are sensible defaults. So I think we should just make that the default, `pebble add <layer_path>`. Can always add `--append` or `--action=append` or something later.
2) For the add algorithm/action, I've gone with `--action=foo` rather than a boolean as there are more than two possibilities, and you don't want multiple booleans with only certain combinations that work. Right now only the default of `--action=combine` is supported, but we could `--action=append` or `--action=insert --index=3` or similar later.
3) I'm not sure we'll need to add static layers, but if we do, I'd prefer `--static` rather than always having to specify `--dynamic` now.

Reply:
Totally agree on point 1. What we need to figure out is what's the best experience we can design for people interacting with the command line. Then, the default should be based on that understanding, and not on what we're doing just now, because we cannot change defaults once this is out. So what I've been trying to imagine is what will someone running a command such as this implicitly imagine it would do:
```
$ pebble add mylayer.yaml
```
The behavior we've defined to start with is a bit quirky to have as a default, unless we polish it a bit further. Maybe we can do that then?

While reading your comments I got another idea that maybe will polish it enough that we can make it common across API, the CLI, and the framework. How about this: let's introduce the label field right now and force people to provide it. Then, the behavior in the document is tuned just slightly to take into account the label. In other words, the possibilies are:

a. Add new layer at the top (order = max order+1), fail if it exists:
    `$ pebble add mylabel mylabel.yaml`

b. Same as before, but combine if it exists:
    `$ pebble add --combine mylabel mylabel.yaml`

c. Same as _a,_ but persist if it exists:
    `$ pebble add --persist mylabel mylabel.yaml`

d. Same as _b_ + _c_:
    `$ pebble add --combine --persist mylabel mylabel.yaml`

For this PR seems closer to _b_, I think?  So maybe we can just support that for now and then walk on with smaller improvements. If that sounds reasonable to you, I will update the document as well.

The --action proposal seems a bit too close to the wire, which is probably comfortable from a development standpoint, but I'm not sure it'd lead us into a nice experience for people using the CLI itself. In fact, even using the the operator framework it might be a bit unfriendly, in the sense of having to figure indexes, for example.

Reply:
Per Mattermost discussion, we're going to go with requiring a label as per above. I'll update the PR to support `pebble add label layer.yaml` and `pebble add --combine label layer.yaml` (but leave out `--persist` and other options for now).

Reply:
This is done now.